src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -106,32 +106,65 @@ std::shared_ptr<RayObject> GetRequest::Get(const ObjectID &object_id) const {
 
 CoreWorkerMemoryStore::CoreWorkerMemoryStore() {}
 
-Status CoreWorkerMemoryStore::Put(const ObjectID &object_id, const RayObject &object) {
-  std::unique_lock<std::mutex> lock(lock_);
-  auto iter = objects_.find(object_id);
-  if (iter != objects_.end()) {
-    return Status::ObjectExists("object already exists in the memory store");
+void CoreWorkerMemoryStore::GetAsync(
+    const ObjectID &object_id, std::function<void(std::shared_ptr<RayObject>)> callback) {
+  std::shared_ptr<RayObject> ptr;
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    auto iter = objects_.find(object_id);
+    if (iter != objects_.end()) {
+      ptr = iter->second;
+    } else {
+      object_async_get_requests_[object_id].push_back(callback);
```

Comment:
Use the same hashmap as `objects_`?

Reply:
I'm not sure how to do this.

Reply:
Ah, I meant to squash `object_async_get_requests_` and `objects_` into one hashmap, so each entry is actually the object value plus any callbacks.

Reply:
I see, I don't think that's worth it right now but maybe if we add more stuff.